### ISSUE TITLE
Denoising Diffusion Implicit Models

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#156]
## TL; DR

- Discover that denoising objective of DDPM is applicable not only to the Markovian process mentioned in the paper but also other non-Markovian processes that match the ***marginals***.
- By introducing a new parameter *\sigma* that controls the stochasticity of the process, authors succeed in controlling the trade off between sample quality and inference time in a favorable direction.

## Motivation

- DDPMs have shown promising performance but are extremely slow for sampling, rendering it inadequate for real-life applications.

## Prior Works

- DDPM (Ho et. al., 2020)

## Method

- Observe that the denoising objective of DDPM can be generally applied to both Markovian and non-Markovian processes that share the same marginal distributions (*not joints, there are many joint distributions that share the same marginals*)
- Hence, define a family of inference distributions (analogous to noising process in DDPM) that is additionally parameterized by *\sigma* that yields the same mean functions, hence the same marginals.
![image](https://user-images.githubusercontent.com/46308407/152951245-629ec239-00ef-485f-a8ab-8073937ea23f.png)
- With Bayes rule, we can define a forward process that is non-Markovian but perfectly suits the *exact same objective* of DDPM.

![image](https://user-images.githubusercontent.com/46308407/152951329-140cd9b2-5139-4176-8f8f-823216caecb8.png)
 
- Now we define a trainable generative process that matches not the Markovian reverse process conditioned by x0 but the non-Markovian process parameterized by \sigma.
- To that end, first define our *prediction on the denoised input x0* as the output of neural network (parameterized by epsilon, the noise term originally predicted by the model)
![image](https://user-images.githubusercontent.com/46308407/152951686-df9b10e4-7676-4e5d-a20d-bc395ea3aefe.png)
- Then, we can define the learnable reverse process as function of x_(t) and predicted x_(0).
- Although our objective below depends on \sigma, with model parameters unshared across t, our \sigma-dependent objective is identical to the vanilla DDPM objective in terms of optimization.
![image](https://user-images.githubusercontent.com/46308407/152951963-a979e7c8-fd89-4911-93fe-c3ce2b8fd652.png)

- Under DDIM framework, our learned reverse process not only depend on x_(t) but also predicted x_(0) like follows:
![image](https://user-images.githubusercontent.com/46308407/152952152-79207b8e-76eb-416f-8d13-231dca0f86c6.png)

- Although skipping denoising steps lead to significant drop in sample quality in vanilla DDPM due to its Markovian nature, the phenomenon is much milder with DDIM as it is no longer Markovian (hence no explicit need to go over the full Markov chain)

- With DDIM, we have two controllable parameters, \tau and \sigma. 
- \tau controls how many steps to go over, and \sigma controls how much stochasiticity to induce in our process.
- With adequate combination of \tau and \sigma, we can achieve favorable trade of between sample quality and inference time (proportional to # steps)
- When \sigma=0, the authors call it DDIM as there is no explicit stochasticity in the process, and the output sample is fully decided by the latent variable (latent variable model).

- (Skipped the part about *relevance to ODEs*)

## Experiments

- Controls \sigma with \eta for simplicity. \eta=1 corresponds to DDPM, and \eta=0 indicates DDIM.

![image](https://user-images.githubusercontent.com/46308407/152952874-b8027dc4-ef27-4431-a406-5b5707dcc5a6.png)

- Favorable sample quality with fewer steps.

![image](https://user-images.githubusercontent.com/46308407/152952922-39e81508-dd67-4b79-b595-30c589be2820.png)

- Inference time can be greatly saved, and since DDIM is deterministic, direct interpolation and reconstruction is possible. (No randomness in the generative/noising process)
![image](https://user-images.githubusercontent.com/46308407/152953019-ebe1937a-2557-47ea-971e-1f383d3e928b.png)
![image](https://user-images.githubusercontent.com/46308407/152953033-ba858354-b911-4322-820b-3be2f12dd3ea.png)

## Comments

- Very difficult to understand when I first read, mostly because the paper is very much conceptual, with little *actual* modifications made to the original paper of DDPM.
- Very clever idea and persuasive theoretical drive. Great paper, I'd say.
